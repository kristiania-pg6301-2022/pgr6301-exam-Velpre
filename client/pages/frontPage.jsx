import { Link } from "react-router-dom";
import React from "react";

export function FrontPage() {
    //skal liste bare titell fra DB med en gang
  return (
    <div>
      <h1>VÃ¥re Artikler</h1>
      <ul>
        <li>
          <Link to={"/movies"}>List movies</Link>
        </li>
      </ul>
    </div>
  );
}
