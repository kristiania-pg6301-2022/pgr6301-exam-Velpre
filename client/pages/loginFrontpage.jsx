import { randomString } from "../lib/randomString";
import { sha256 } from "../lib/sha256";
import React, { useContext, useEffect, useState } from "react";
import {Route, Routes, useNavigate, useParams} from "react-router-dom";
import { ApplicationContext } from "../applicationContext";

function LoginButton({ config, label, provider }) {
    async function handleLogin() {
        const {
            authorization_endpoint,
            response_type,
            scope,
            client_id,
            code_challenge_method,
        } = config[provider];

        const state = randomString(50);
        window.sessionStorage.setItem("expected_state", state);

        const parameters = {
            response_type,
            response_mode: "fragment",
            client_id,
            state,
            scope,
            redirect_uri: `${window.location.origin}/login/${provider}/callback`,
        };

        if (code_challenge_method) {
            const code_verifier = randomString(50);
            window.sessionStorage.setItem("code_verifier", code_verifier);
            parameters.code_challenge_method = code_challenge_method;
            parameters.code_challenge = await sha256(code_verifier);
            parameters.domain_hint = "egms.no";
        }

        window.location.href =
            authorization_endpoint + "?" + new URLSearchParams(parameters);
    }

    return (
        <div>
            <button onClick={handleLogin}>{label}</button>
        </div>
    );
}

export function LoginCallback({ reload, config }) {
    const { provider } = useParams();
    const [error, setError] = useState();
    const navigate = useNavigate();
    const { registerLogin } = useContext(ApplicationContext);
    useEffect(async () => {
        const { access_token, error, error_description, state, code } =
            Object.fromEntries(
                new URLSearchParams(window.location.hash.substring(1))
            );

        const expected_state = window.sessionStorage.getItem("expected_state");
        if (!state || expected_state !== state) {
            setError("Unexpected state");
            return;
        }

        if (error || error_description) {
            setError(`Error: ${error} (${error_description})`);
            return;
        }

        if (code) {
            const { client_id, token_endpoint } = config[provider];
            const code_verifier = window.sessionStorage.getItem("code_verifier");
            const redirect_uri = `${window.location.origin}/login/${provider}/callback`;
            //Denne m√• kanskje endres
            const payload = {
                grant_type: "authorization_code",
                code,
                client_id,
                code_verifier,
                redirect_uri,
            };
            const res = await fetch(token_endpoint, {
                method: "POST",
                body: new URLSearchParams(payload),
            });
            if (!res.ok) {
                setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
                return;
            }
            const { access_token } = await res.json();
            await registerLogin(provider, { access_token });
            reload();
            navigate("/front-db");
            return;
        }

        if (!access_token) {
            setError("Missing access_token");
            return;
        }

        await registerLogin(provider, { access_token });
        reload();
        navigate("/front-db");
    }, []);

    if (error) {
        return (
            <div>
                <h1>Error</h1>
                <div>{error.toString()}</div>
            </div>
        );
    }

    return <h1>Please wait...</h1>;
}


function StartLogin({ config }) {
    return (
        <div>
            <h1>Login</h1>
            <LoginButton
                label={"Login with Google"}
                config={config}
                provider={"google"}
            />
            <LoginButton
                label={"Login with ID-porten"}
                config={config}
                provider={"hk"}
            />
        </div>
    );
}

export function LoginFrontPage({ config, reload }) {
    return (
        <Routes>
            <Route path={"/"} element={<StartLogin config={config} />} />
            <Route
                path={"/:provider/callback"}
                element={<LoginCallback config={config} reload={reload} />}
            />
            <Route path={"*"} element={<StartLogin config={config} />} />
        </Routes>
    );
}

//<Route path={"/endsession"} element={<EndSession reload={reload} />} /> - TRENGER VI DENNE
