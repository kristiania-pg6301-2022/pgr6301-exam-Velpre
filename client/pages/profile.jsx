import React from "react";
import { useNavigate } from "react-router-dom";

export function Profile({ user, reload, logout }) {

  const navigate = useNavigate();

  async function handleLogout() {
    await logout("/api/login", reload, navigate);
  }

  return (
    <div className="profile">
        <h1>Profile</h1>
      {user.google ? (
        <div>
          <h1>
            User profile: {user.google.name} ({user.google.email})
          </h1>
          <img
            src={user.google.picture}
            alt={user.google.name + " profile picture"}
          />
        </div>
      ) : (
        <h1>
          User profile: {user.hk.name} ({user.hk.email})
        </h1>
      )}
      <div>
        <button className="btn" onClick={handleLogout}>Log out</button>
      </div>
    </div>
  );
}
