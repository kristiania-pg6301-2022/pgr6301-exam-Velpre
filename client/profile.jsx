import React, { useContext } from "react";
import { ApplicationContext } from "./applicationContext";
import { useNavigate } from "react-router-dom";

export function Profile({ user, reload }) {
  const { logout } = useContext(ApplicationContext);
  const navigate = useNavigate();

  async function handleLogout() {
    await logout("/api/login", reload, navigate);
  }

  return (
    <div>
      {user.google ? (
        <div>
          <h1>
            User profile: {user.google.name} ({user.google.email})
          </h1>
          <img
            src={user.google.picture}
            alt={user.google.name + " profile picture"}
          />
        </div>
      ) : (
        <h1>
          User profile: {user.hk.name} ({user.hk.email})
        </h1>
      )}
      <div>
        <button onClick={handleLogout}>Log out</button>
      </div>
    </div>
  );
}
