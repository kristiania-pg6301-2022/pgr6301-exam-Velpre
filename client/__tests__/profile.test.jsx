import * as React from "react";
import { render } from "react-dom";
import { MemoryRouter } from "react-router-dom";
import { Profile } from "../pages/profile";
import { Simulate } from "react-dom/test-utils";

describe("Profile compoent", () => {
  it("Render users", () => {
    const user = {
      google: {
        email: "sajmon.vp@gmail.com",
        name: "V P",
        picture:
          "https://lh3.googleusercontent.com/a/AATXAJzKiTXm2d-4PFX3RwEkS6MBhjOGZNv4PBwEQ3G-=s96-c",
      },
    };
    const fun = jest.fn();
    const element = document.createElement("div");
    render(
      <MemoryRouter initialEntries={["/profile"]}>
        <Profile user={user} reload={fun} />
      </MemoryRouter>,
      element
    );

    expect(element.querySelector("h1").innerHTML).toEqual("You are not loged in");
    expect(element.innerHTML).toMatchSnapshot();
  });
});
