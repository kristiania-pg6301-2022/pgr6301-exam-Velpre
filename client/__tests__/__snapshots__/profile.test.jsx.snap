// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile compoent Render users 1`] = `"<div class=\\"profile\\"><h1>You are not loged in</h1><div><h1>User profile: V P (sajmon.vp@gmail.com)</h1><img src=\\"https://lh3.googleusercontent.com/a/AATXAJzKiTXm2d-4PFX3RwEkS6MBhjOGZNv4PBwEQ3G-=s96-c\\" alt=\\"V P profile picture\\"></div><div><button class=\\"btn\\">Back</button></div></div>"`;
